### ISSUE TITLE
Talk about citre-tree, a "git-like" code reading tool

### ISSUE THREAD

**Issue by AmaiKinono** [AmaiKinono/citre#20]
This is a sub-issue of #17.

This is what @masatake has mentioned in our private mails. The idea is to keep the code reading history as a tree, and we could do git-like operations on it, like changing branch, and maybe rebase. Here's his summary:

> What I want may be a "git-for-code-readers". The tool can record the process of code-reading as commits. The chain of commits can be edited and refined. The final version of the chain is something like a proof in mathematics. The tools manages multipe chains as branches.

We both don't have a clear concept of what this tool will be like, or @masatake actually has one and he didn't tell me ;) Anyway, let's call it `citre-tree` for now.

## Maybe `citre-map` could do the job.

First of all, `citre-map` already gives us a tree that can be edited. The difference between that and `citre-tree` is, `citre-map` draws the tree based on the relationship between the items (file -> symbol -> definition), and not the history of reading, while `citre-tree` may do the latter. I think they are trying to solve the same problem: `citre-map` also lets you browse in the codebase broadly, then cut the branches that's not important, and eventually leaves a small map that solves the problem. So, if `citre-map` does this well, we won't need `citre-tree`.

## Why `citre-tree` shouldn't be a part of Citre

Even if `citre-map` is not good, and we need `citre-tree`, I feel it's not in the scope of this package. Let me explain: I assume in `citre-tree`, you could use current locations as nodes, and put them into a tree. Since the reading process is non-linear, it doesn't mean locations you jump to later appears deeper in the tree. Actually the relationship between a parent node A and its child B could be:

- B is the definition of A. When we chasing down a calling tree, we'll have this.

- B is a reference of A. When we want to see all references of a symbol, we'll have this.

  Notice that we may jump from B to A using "jump to definition", but still want to put B under A. So we don't always jump from the parent node to its child.

- B is used in the body of A.

- Other logical relations.

So, the tree is formed in an logical order, where the logic is neither time order, nor you jump from the parent to the child using our tools, but something that's needed to be figured out by the reader himself. If we could agree this, what we need is a tool that allows us to:

- Collect locations as nodes.
- Link the nodes into little chains, and eventually into a tree (or, a directed graph since there may also be circles in the graph).
- Edit the graph.
- Annotate on the nodes, maybe also on the arrows.

And this actually doesn't has much to do with "jumping to definition/reference" tools. Yes you'll need them, but what they do is just take you from one location to another. You still have to manually create the tree. The tools must not create/modify the tree since they doesn't understand logical relations between the nodes, and only the code reader can figure them out.

If we could agree with this, then I would say it's better to develop it as a new package, so other users could use it with their favorite "jumping to" tools.

**Comment by AmaiKinono** [comment]
I actually find [org-roam](https://github.com/org-roam/org-roam) somewhat resembles the idea of citre-tree. It's impressive that it could generate a graph for all the connected links, and could be used to jump to the files. I think what citre-tree need is just this small part, replacing the "links" by "locations", plus graph editing and annotating functionalities.

**Comment by AmaiKinono** [comment]
I think my recent experience on reading the code of u-ctags (<https://github.com/AmaiKinono/citre/issues/27#issuecomment-723455882>) is helping me to understand a bit of the process of long code reading sessions. I have some thoughts on design such a tool (separate from the Citre project):

- It's a tool that helps you taking notes while reading.

- It's a minor mode that works in any text file, so you could still use your favorite markup language.

- It offers a markup (I think `#[[title]]` is actually good) that we'll highlight using overlays. Your note will look something like:

  ```
  #[[entry.c: makePatternString]] calls #[[entry.c: makePatternStringCommon]], which does a, b and c...
  ```

  We'll offer user options so you can use green for definition, blue for reference, etc.. and it will be customizable.

- These links actually points to positions in files. We'll design some helpers so that you could "copy" a location while reading, and "paste" it in your note, and give it a title.

- The relation between link titles and their locations are stored in a lisp object, so they can be written to a database. For example, if the note file is `note.md`, we could use `.note.md.treedb` as the name of the database. So you can read and modify your note later.

- We'll offer helpers for inserting a link, renaming a link, updating its location, etc..

On Citre's side, I'd like to let `citre-peek` work with this system. Actually we don't have to integrate them. What I want to do is:

- Keep a full history of jumping in `citre-peek` sessions. Every new "jumping from here to here" event will be appended to the history, including the "from" and "to" locations.

- Offer a command for browsing the history and deleting items in it.

So, the code reading process will be:

- We jump to places using `citre-peek`.
- We find something interesting, so we look at the history, find relevant locations and jump there, copy them and paste into our note.
- Since relevant locations are made persistent in the note, we could delete irrelevant history from time to time, making it easier to continue reading.

In summary, we can read code using three things:

- Tool to go to places (`citre-peek` here. Of course you could add `grep` also, this system is not dependent on certain tools.)
- A bucket to record places we went to, so we could rummage through them at any time. (Here it's the history of `citre-peek`.)
- A note-taking tool to put relevant places together, reveal the relationship between them and tell a story.

I feel it's almost like cracking a criminal case (I bet Masatake have watched "Furuhata Ninzaburoo", I like it ;), we go to places, collecting clues, ane connecting relevant ones in our notebook.

**Comment by AmaiKinono** [comment]
I've put the ideas into a prototype called "clue": https://github.com/AmaiKinono/clue

See it in action:

![clue](https://user-images.githubusercontent.com/28714352/99285983-35915880-2873-11eb-85bd-cabab2927690.gif)

I plan to remove `citre-map`, and let `clue` serve for long code reading sessions. @masatake could you try it and tell me if you think it will work?

**Comment by masatake** [comment]
I tried clue. Allow me to write random notes.

How do you call the `#[here]`? I will call `code-link` here.
I wonder what should be included in a `code-link`?
A. source file name?
B. line or range in the file?
C. git commit id?
D. name picked up from tags file and its parent*s* up to the root scope
E. whole contents of the source file?
F. whole source tree (if we have unlimited computing resources)?
...

I would like to use the citre's peek window alike tool to show the detail of `code-link


I agree with you that clue is an important building block for code readers.

However, we need one more tool.
What I need is the upper part of your screen capture: .`"my code reading note'.
I wonder how to manage it.

```
[reading session] = a directory?
`- args.tags file using for running ctags (optional)
`- tags file or link to tags file
`- reading-note.rst
```

I wonder *Bookmark list*(\C-x r l) can be used for managing sessions.

**Comment by AmaiKinono** [comment]
> I wonder what should be included in a `code-link`?

If it's acceptable to use one line per link, I'd like to put the file name and line number in the link title (as I write in our private email). The benefits are:

- The note is readable and meaningful as plain text.
- No need to use a separate database to store link locations, as the location is inside the link title.

This tool will be independent from the "finding definition" tools we use (e.g., you can also use grep), so there won't be ctags related info in the links. If you need more information, you can:

- Write them down in the note,
- or just jump to the location and study it.

> I would like to use the citre's peek window alike tool to show the detail of `code-link`

It's easy to do. The task is similar to #51.

> What I need is the upper part of your screen capture: "my code reading note".
> I wonder how to manage it.

As mentioned above, my plan is to make the note a simple plain text file (you can use whatever markup language you like for it).

You'll need the source tree to jump from the link to the location, but the note itself is readable, and can be posted to the internet or share with other people, even the reader doesn't use Emacs.

> git commit id?

I think we can use "meta" links for these info. Like:

```
#[:meta:git-commit:123456]
#[:meta:project-root:/path/to/project/root]
```

In this way, we can use relative path in the links, and when the project dir changes, we just fix the `project-root` meta link.

For now I can't think of an efficient way to use the git commit.

**Comment by AmaiKinono** [comment]
Close this as the basic ideas are implemented in `citre-peek` and `clue`.
